### DISCUSSION TITLE
Report

### DISCUSSION THREAD

**Discussion by ferraridavide** [ferraridavide/sunomi#33] [General]
Cose da considerare scrivendo il report

- Usiamo le S3 Event Notifications per far partire il flusso di upload quando un video viene caricato
  - Stando alla [documentazione AWS](https://docs.aws.amazon.com/AmazonS3/latest/userguide/EventNotifications.html), questi eventi garanzia di consegna "at least once", normalmente nell'arco di qualche secondo, ma alcune volte possono volerci fino a un minuto o più
      > Amazon S3 event notifications are designed to be delivered at least once. Typically, event notifications are delivered in seconds but can sometimes take a minute or longer.
  - Utilizziamo dei filtri sugli oggetti che triggerano queste notifiche, questi filtri si basano sul tipo di evento (creare, eliminare, ect) e solo su prefissi e suffissi dell'oggetto in questione, questi prefissi e suffissi hanno però la limitazione di [non poter contenere caratteri wildcard](https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-how-to-filtering.html)
      > A wildcard character ("*") can't be used in filters as a prefix or suffix. If your prefix or suffix contains a space, you must replace it with the "+" character. If you use any other special characters in the value of the prefix or suffix, you must enter them in URL-encoded (percent-encoded) format. For a complete list of special characters that must be converted to URL-encoded format when used in a prefix or suffix for event notifications, see Safe characters.
      
- Per notificare il frontend sullo stato di elaborazione del video in upload abbiamo inizialmente (nella fase di sviluppo in locale) utilizzato gli SSE (Server-Sent Events), adatti in questo caso in quanto siamo interessati ad una comunicazione unilaterale da server a client
  - Abbiamo riscontrato che, pur essendo [supportata in tutti i browser moderni](https://caniuse.com/eventsource), l'implementazione standard definita dal [W3C HTTP Standard](https://html.spec.whatwg.org/multipage/server-sent-events.html) {se menzioniamo questo + il link siamo dei campioni} dell'API EventSource è molto limitata e manca di funzionalità importanti, non permette nativamente di passare headers o altri parametri di connessione, per queste funzionalità, siamo dovuti ricorrere all'utilizzo di [librerie polyfill](https://github.com/Azure/fetch-event-source)
  

    > The [default browser EventSource API](https://developer.mozilla.org/en-US/docs/Web/API/EventSource) imposes several restrictions on the type of request you're allowed to make: the [only parameters](https://developer.mozilla.org/en-US/docs/Web/API/EventSource/EventSource#Parameters) you're allowed to pass in are the url and withCredentials, so:
    > 
    > - You cannot pass in a request body: you have to encode all the information necessary to execute the request inside the URL, which is [limited to 2000 characters](https://stackoverflow.com/questions/417142) in most browsers.
    > - You cannot pass in custom request headers
    > - You can only make GET requests - there is no way to specify another method.
    > - If the connection is cut, you don't have any control over the retry strategy: the browser will silently retry for you a few times and then stop, which is not good enough for any sort of robust application.

   - Successivamente, passando alla progettazione dell'architettura cloud su AWS, viste le necessità di dover spostare la gestione di queste connessioni al di fuori dei "controller" (in quanto questi devono scalare, e le connessioni aperte potrebbero venir chiuse in un processo di scale down), abbiamo deciso di utilizzare AWS API Gateway e i WebSocket
     - Prima di spostare questa funzionalità dal "controller" a API Gateway, le modifiche lato controller e frontend per passare da SSE a WebSocket sono state minime, essenzialmente un drop-in replacement
     - Spiegare il flusso con route `$connect`, `$disconnect`, collegate a Lambda e DynamoDB, spiegare perchè DynamoDB e possibile utilizzo di AWS ElastiCache e AWS MemoryDB
 
 
 - Transcoder
    - Spiegare la decisione di processare tutte le qualità + thumbnail dei video in un singolo processo transcoder, piuttosto che processare ogni qualità in parallelo su processi diversi (numero di worker, bandwidth, considerazioni sul numero di download di file, ect)
    - Spiegare la decisione di utilizzare un transcoder interno piuttosto che AWS Elemental MediaConvert (vedi #10)
    
    
 - Sistema generale
   - Spiegare la scelta di aspettare/non aspettare che ogni processo tra transcoding, trascrizione e rilevazione di contenuti sensibili sia terminato prima di rendere il video pubblico 
   
   
   
 Problematiche affrontate:
  - La porta 3000 bloccata dal Wifi dell'università
   - Attenzione ai timeout di default delle lambda, di default una lambda ha un timeout di 3 secondi, questo ci ha portato spesso a problemi per cui una lambda viene uccisa e poi riprovata (di default c'è anche un retry attempts = 2), questo ci portava ad avere un problema per cui la lambda che triggera l'avvio del transcoder veniva raramente eseguita due volte (la prima volta uccisa per timeout (forse a causa di cold start) ma con lavoro utile fatto, la seconda normalmente ma avviata per via del retry)

**Comment by alerotta** [comment]
- S3 Bucket

**ACL disabilitate (raccomandato da AWS):**
Gli oggetti all'interno del bucket sono di proprietà esclusiva dell'account AWS che li crea. L'accesso al bucket e al suo contenuto è gestito esclusivamente tramite le policies.

**Accesso pubblico disabilitato (raccomandato da AWS):**
L'accesso al bucket è controllato interamente tramite policies, garantendo maggiore sicurezza.

- CloudFront Distribution (CDN)

**Origin:** S3 Bucket
**Origin Shield:** No
**Connection Retry Setting:** Default
**Protocollo Utente:** HTTP, HTTPS o entrambi
**Metodi Utente:** GET e HEAD (solo per la visualizzazione dei contenuti)
**Restrict Viewer Access:** No (non è richiesto un presigned URL o cookie per accedere ai contenuti)
**Cache Policy:** Da definire (esistono preset disponibili)
**Origin Request Policy:** Da definire
**Response Header Policy (opzionale):** Da definire
**Functions Association:** Nessuna
**Firewall Security:** Abilitata
**Anycast IP List:** Necessaria e da creare
**Price Class:** Possibilità di scegliere tra tutte le regioni o solo quelle principali
**Alternative Domain Name:** No
**Custom SSL:** No
**Supported HTTP Versions:** Scegliere se supportare HTTP/2 e HTTP/3 (HTTP/1 e 1.1 sono supportati di default)
**IPv6**: ?
**Logging:** Default

**Reply by Fede-droid** [reply]
CDN - VIDEO: url statico o signed?

**Reply by Fede-droid** [reply]
non abbiamo aggiunto il WAF. abbiamo messo WAF: "disabilited"
